test/unit/sdam/server_discovery_and_monitoring.prose.test.ts

```diff
@@ -0,0 +1,63 @@
+import { expect } from 'chai';
+import { createServer, type Server } from 'net';
+
+import { MongoClient, SERVER_HEARTBEAT_FAILED, SERVER_HEARTBEAT_STARTED } from '../../mongodb';
+
+describe('Heartbeat tests', function () {
+  let client: MongoClient;
+  // Mock server
+  let server: Server;
+  // Shared array
+  const events: string[] = [];
+
+  beforeEach(async function () {
+    // Create TCP server that responds to hellos by closing the connection
+    // and pushing "client connection created" to shared array
+    server = createServer(clientSocket => {
+      events.push('client connected');
+
+      clientSocket.once('data', () => {
+        events.push('client hello received');
+        clientSocket.end();
+      });
+    });
+    server.listen(9999);
```

Comment:
Can we make a promise out of the [`'listening'`](https://nodejs.org/docs/latest/api/net.html#event-listening) event and await it? I think we're "getting lucky" that the server has bound to the 9999 port because we await client.connect()